consensus/src/consensus_observer/pending_blocks.rs

```diff
@@ -5,98 +5,205 @@ use crate::consensus_observer::{
-    /// This will remove all blocks up to (and including) the commit
-    /// round of the committed ledger info.
+    /// Removes the pending blocks for the given commit ledger info. This will
+    /// remove all blocks up to (and including) the epoch and round of the
+    /// commit. Note: this function must remove both verified and unverified
+    /// blocks (to support state sync commits).
     pub fn remove_blocks_for_commit(&self, commit_ledger_info: &LedgerInfoWithSignatures) {
-        // Determine the round to split off
-        let split_off_round = commit_ledger_info.commit_info().round() + 1;
-
-        // Remove the pending blocks before the split off round
-        let mut pending_ordered_blocks = self.pending_ordered_blocks.lock();
-        *pending_ordered_blocks = pending_ordered_blocks.split_off(&split_off_round);
+        // Determine the epoch and round to split off
+        let split_off_epoch = commit_ledger_info.ledger_info().epoch();
+        let split_off_round = commit_ledger_info.commit_info().round().saturating_add(1);
+
+        // Remove the blocks from the verified pending blocks
+        let mut verified_pending_blocks = self.verified_pending_blocks.lock();
+        *verified_pending_blocks =
+            verified_pending_blocks.split_off(&(split_off_epoch, split_off_round));
+
+        // Remove the blocks from the unverified pending blocks
+        let mut unverified_pending_blocks = self.unverified_pending_blocks.lock();
+        *unverified_pending_blocks =
+            unverified_pending_blocks.split_off(&(split_off_epoch, split_off_round));
     }
 
-    /// Updates the commit decision of the pending ordered block (if found)
+    /// Updates the commit decision of the pending ordered block (if found).
+    /// This can only be done for verified pending blocks.
     pub fn update_commit_decision(&self, commit_decision: &CommitDecision) {
-        let mut pending_ordered_blocks = self.pending_ordered_blocks.lock();
+        // Get the epoch and round of the commit decision
+        let commit_decision_epoch = commit_decision.epoch();
+        let commit_decision_round = commit_decision.round();
+
+        // Update the commit decision for the verified pending blocks
+        let mut verified_pending_blocks = self.verified_pending_blocks.lock();
         if let Some((_, existing_commit_decision)) =
-            pending_ordered_blocks.get_mut(&commit_decision.round())
+            verified_pending_blocks.get_mut(&(commit_decision_epoch, commit_decision_round))
         {
             *existing_commit_decision = Some(commit_decision.clone());
         }
     }
-}
 
-impl Default for PendingOrderedBlocks {
-    fn default() -> Self {
-        Self::new()
+    /// Verifies the pending blocks against the given epoch state.
+    /// If verification is successful, blocks are marked as verified.
+    pub fn verify_pending_blocks(&self, epoch_state: &EpochState) {
```

Comment:
Convince me that it's okay not to have a single lock around both .unverified_pending_blocks and .verified_pending_blocks ? Access order makes this okay? Comment here on how we are sure this is right?

Reply:
There is a possibility of commiut if `remove_blocks_for_commit` and this method is called concurrently.

Reply:
Yeah, this was all cleaned up in https://github.com/aptos-labs/aptos-core/pull/13930/files, i.e.,:
- Don't ever hold both locks at once.
- If we must, ensure the same lock ordering.

I also have another pending PR which goes through and avoids both locks (as @ibalajiarun suggested below), but I'll redo this PR and just merge it all together 😄